ethers/providers/jsonrpc/subscriptions.nim

```diff
@@ -137,34 +167,69 @@ method subscribeLogs(subscriptions: WebSocketSubscriptions,
     let res = Log.fromJson(arguments{"result"}).mapFailure(SubscriptionError)
     onLog(res)
 
-  convertErrorsToSubscriptionError:
-    let id = await subscriptions.client.eth_subscribe("logs", filter)
-    subscriptions.callbacks[id] = callback
-    return id
+  try:
+    withLock(subscriptions):
+      convertErrorsToSubscriptionError:
+        let id = await subscriptions.client.eth_subscribe("logs", filter)
+        subscriptions.callbacks[id] = callback
+        subscriptions.logFilters[id] = filter
+        return id
+  except AsyncLockError as e:
+     error "Lock error when trying to subscribe to logs", err = e.msg
+     raise newException(SubscriptionError, "Cannot subscribe to the logs because of lock error")
 
 method unsubscribe*(subscriptions: WebSocketSubscriptions,
                    id: JsonNode)
                   {.async: (raises: [CancelledError]).} =
   try:
-    subscriptions.callbacks.del(id)
-    discard await subscriptions.client.eth_unsubscribe(id)
+    withLock(subscriptions):
+      subscriptions.callbacks.del(id)
+      discard await subscriptions.client.eth_unsubscribe(id)
   except CancelledError as e:
     raise e
   except CatchableError:
     # Ignore if uninstallation of the subscribiton fails.
     discard
 
+method close*(subscriptions: WebsocketSubscriptions) {.async.} =
+  await procCall JsonRpcSubscriptions(subscriptions).close()
+  when defined(ws_resubscribe):
+    if not subscriptions.resubscribeFut.isNil:
+        await subscriptions.resubscribeFut.cancelAndWait()
+
+when defined(ws_resubscribe):
+  # This is a workaround to manage the 5 minutes limit due to hardhat.
+  # See https://github.com/NomicFoundation/hardhat/issues/2053#issuecomment-1061374064
+  proc resubscribeWebsocketEventsOnTimeout*(subscriptions: WebsocketSubscriptions) {.async: (raises: [CancelledError]).} =
+    while true:
```

Comment:
Maybe we should do a bit of validation here, returning if the resubscribe interval is <= 0?